winpr/libwinpr/sspi/Kerberos/kerberos.c

```diff
@@ -659,117 +659,196 @@ static BOOL kerberos_mk_tgt_token(SecBuffer* buf, int msg_type, char* sname, cha
+		}
+	}
 
-		buf = malloc(len);
-		if (!buf)
-			return FALSE;
+	if (checkForTag3)
+	{
+		if (tag == 3)
+			rc = kerberos_rd_tgt_req_tag3(&dec2, buf, len, &offset);
+		else
+			rc = FALSE;
+	}
 
-		*buf = 0;
+end:
+	if (rc)
 		*target = buf;
+	else
+		free(buf);
+	return rc;
+}
 
-		if (!WinPrAsn1DecReadContextualTag(&dec, &tag, &dec2))
-			goto fail;
+static BOOL kerberos_rd_tgt_rep(WinPrAsn1Decoder* dec, krb5_data* ticket)
+{
+	if (!ticket)
+		return FALSE;
 
-		if (tag == 2)
-		{
-			WinPrAsn1Decoder seq;
-			/* server-name [2] PrincipalName (SEQUENCE) */
-			if (!WinPrAsn1DecReadSequence(&dec2, &seq))
-				goto fail;
+	/* ticket [2] Ticket */
+	WinPrAsn1Decoder asnTicket = { 0 };
+	WinPrAsn1_tagId tag = 0;
+	if (WinPrAsn1DecReadContextualTag(dec, &tag, &asnTicket) == 0)
+		return FALSE;
 
-			/* name-type [0] INTEGER */
-			if (!WinPrAsn1DecReadContextualInteger(&seq, 0, &error, &val))
-				goto fail;
+	if (tag != 2)
+		return FALSE;
 
-			/* name-string [1] SEQUENCE OF GeneralString */
-			if (!WinPrAsn1DecReadContextualSequence(&seq, 1, &error, &dec2))
-				goto fail;
+	wStream s = WinPrAsn1DecGetStream(&asnTicket);
+	ticket->data = Stream_BufferAs(&s, char);
+	ticket->length = Stream_Length(&s);
+	return TRUE;
+}
 
-			while (WinPrAsn1DecPeekTag(&dec2, &tag))
-			{
-				if (!WinPrAsn1DecReadGeneralString(&dec2, &str))
-					goto fail;
+static BOOL kerberos_rd_tgt_token(const sspi_gss_data* token, char** target, krb5_data* ticket)
+{
+	BOOL error = 0;
+	WinPrAsn1_INTEGER val = 0;
 
-				if (buf != *target)
-					*buf++ = '/';
-				buf = stpcpy(buf, str);
-				free(str);
-			}
+	WINPR_ASSERT(token);
 
-			if (!WinPrAsn1DecReadContextualTag(&dec, &tag, &dec2))
-				return TRUE;
-		}
+	if (target)
+		*target = NULL;
 
-		/* realm [3] Realm */
-		if (tag != 3 || !WinPrAsn1DecReadGeneralString(&dec2, &str))
-			goto fail;
+	WinPrAsn1Decoder der = { 0 };
+	WinPrAsn1Decoder_InitMem(&der, WINPR_ASN1_DER, (BYTE*)token->data, token->length);
 
-		*buf++ = '@';
-		strcpy(buf, str);
-		free(str);
-		return TRUE;
-	}
-	else if (val == KRB_TGT_REP)
-	{
-		if (!ticket)
-			return FALSE;
+	/* KERB-TGT-REQUEST (SEQUENCE) */
+	WinPrAsn1Decoder seq = { 0 };
+	if (!WinPrAsn1DecReadSequence(&der, &seq))
+		return FALSE;
 
-		/* ticket [2] Ticket */
-		if (!WinPrAsn1DecReadContextualTag(&dec, &tag, &dec2) || tag != 2)
-			return FALSE;
+	/* pvno [0] INTEGER */
+	if (!WinPrAsn1DecReadContextualInteger(&seq, 0, &error, &val) || val != 5)
+		return FALSE;
 
-		s = WinPrAsn1DecGetStream(&dec2);
-		ticket->data = Stream_BufferAs(&s, char);
-		ticket->length = Stream_Length(&s);
-		return TRUE;
-	}
-	else
+	/* msg-type [1] INTEGER */
+	if (!WinPrAsn1DecReadContextualInteger(&seq, 1, &error, &val))
 		return FALSE;
 
-fail:
-	free(buf);
-	if (target)
-		*target = NULL;
-	return FALSE;
+	switch (val)
+	{
+		case KRB_TGT_REQ:
+			return kerberos_rd_tgt_req(&seq, target);
+		case KRB_TGT_REP:
+			return kerberos_rd_tgt_rep(&seq, ticket);
+		default:
+			return FALSE;
```

Comment:
Always a friend of breaking in default and have a final `return FALSE` at the end of the function to avoid future issues